Add docker

alekseytumbaev opened the pull request.

alekseytumbaev committed:
feat:
- gateway and server modules

refactor:
- substitute ItemNullFieldsException and UserNullFieldsException with Creation constraint group

test:
- move tests with validation to gateway

alekseytumbaev committed:
feat:
- Containers for gateway, server and db with debugging ports
- Versions for some dependencies in poms

delete:
- UnknownBookingStateException from server, because booking state is checked by gateway
- Test of UnknownBookingStateException in server
- mvnw and add maven-wrapper to gitignore to pass no binary files test

test:
- Test for UserNotFoundException in UserControllerTest
- test.properties in gateway test/resources

chehovnp commented:
Обрати внимание, что много коммитов висит в гите. Чаще всего в работе на один МР один коммит.
Это добивается ведением одного коммита в процессе разработки фичи "amend commit"/"force push" (при условии, что разработка ведется одним программистом) или слиянием коммитов в один перед МР "git squash". 

В пом можно использовать блок пропертис для обозначении там версий библиотек. В целом, вижу что ты разобрался и выполнил работу в соответствии с ТЗ, имеет смысл подключить remote debug к обоим приложениям в докере в плане тренировки и обучения. Что еще можно: ТЗ намекает на, то что шлюз можно запустить в нескольких экземплярах. Это подразумевает использования сервера haproxy и разделение нагрузки между экземплярами. Есть время и желание , то почитай , можно в компоузе дописать конфигурацию и получить high availability приложение . Что еще можно сделать лучше : файл readme не информативный, нужно сделать описание: что за приложение, как взаимодействует, как его собирать - запускать , как конфигурировать.

По работе важных замечаний нет. Работу принимаю.  Если что-то дополнишь и нужно ответить или посмотреть еще раз, то пиши в личку.

alekseytumbaev committed:
feat:
- haproxy, create 2 instances of gateway

refactor:
- add logging responses in gateway controllers
- BaseClient now creates new ResponseEntity out of server's response, because haproxy will send 502 if not do it

delete:
- remove debugging port from gateway

style:
- rename ObjectValidationViolation.name to message

test:
- add postman-tests.json

docs:
add endpoints description to README
